### PR TITLE
[MRG] refactor `gather` functionality for speed & modularity; provide `prefetch` functionality.

### LINKED ISSUE #1310: maybe add new command, `sourmash prefetch`
I'm a big fan of the [`prefetch_gather` functionality in genome-grist](https://github.com/dib-lab/genome-grist/blob/latest/genome_grist/prefetch_gather.py) that pulls out overlapping signatures from a large database for the purpose of doing downstream gather and other things, and it seems like https://github.com/dib-lab/sourmash/pull/1238 is going to be adding the related functionality from greyhound #1226 into sourmash. However, I am wondering if there's value in explicitly adding a Python-based sourmash CLI function, `sourmash prefetch`, that would let us explore the behavior in non-Rust code. It seems like it would be generically useful to have it be available, and it would also prevent us from copy-pasting `prefetch_gather` over to other projects like charcoal and maybe thumper...

### PR DESCRIPTION
**Pull request by ctb**
Fixes #1310.

This PR revamps the overall `gather` functionality in sourmash to be much more flexible and approximately 50% faster based on [benchmarking](https://github.com/dib-lab/sourmash/pull/1370#issuecomment-833944673) as well as (much) less memory intensive. The ultimate goal is to support oxidation of `gather` per #1226, but this also adds a lot of nice new functionality. 

Changes and additions:

* refactors `gather_databases` to use a new `CounterGather` interface that supports optimized `gather` functionality with `peek` and `consume`; these objects are returned by a new `Index.counter_gather(...)` method. See #1370 for detailed discussion and benchmarking.
* adds a `prefetch` command-line subcommand that does a streaming pass across the provided databases, doing so either with a (potentially optimized) `Index.find(...)` call from #1392, or with a linear pass across all signatures yielded by `Index.signatures()`.
* refactors `Index.gather(...)` functionality to use `Index.prefetch(...)` underneath by default;
* tests all `sourmash gather` commands with all four combinations of prefetch and linear.
* provides a `--save-prefetch` option on `sourmash gather` to save all of the overlapping signatures before digesting them down to the min set cover.
* creates a new `LazyLinearIndex(...)` class that defers signature selection until as late as possible, to support better prefetch.
* Adds a test for the (previously untested) `sourmash search --save-matches`
* change `Index.find(...)` to return an `IndexSearchResult` to preserve location
* add `.signatures_with_location()` to `Index` base class, so as to preserve location in base `Index.find(...)` functionality

Note: this PR includes the new `--save-matches` behavior from https://github.com/dib-lab/sourmash/pull/1493.

## Motivation, inspiration, and related issues

The prefetch functionality is inspired by`prefetch_gather` in genome-grist as well as greyhound #1226.

This PR makes it fairly straightforward to support reporting of all matches with equal containment for large databases, which can help address https://github.com/dib-lab/sourmash/issues/278 https://github.com/dib-lab/sourmash/issues/707 https://github.com/dib-lab/sourmash/issues/1366 https://github.com/dib-lab/sourmash/issues/1198

In particular, we can now support better tie-breaking in gather-style searching https://github.com/dib-lab/sourmash/issues/1366, as well as "reverse" gather https://github.com/dib-lab/sourmash/issues/1198

## CounterGather interface

A key addition in this PR is the addition and systematic usage of a `CounterGather` interface, based on @luizirber PR https://github.com/dib-lab/sourmash/pull/1311, which in turn is motivated by [the greyhound issue](https://github.com/dib-lab/sourmash/issues/1226) and [pull request](https://github.com/dib-lab/sourmash/pull/1238).

The core code is located in `src/sourmash/index.py`, which provides a `CounterGather` class that collects and prioritizes matches for gather and also supports cross-database gather. This object is a query-specific object created and returned by the `Index.counter_gather(...)` method which (by default) uses `Index.prefetch(...)` underneath to do a single pass across the database and collect all possible matches.

The key piece of the refactoring is that `CounterGather` now provides two methods, `peek(query)` and `consume(...)`. `peek(query)` provides the best containment result from _this_ counter, but does not adjust any of the internal information; `consume(...)` is used to remove a match (potentially from _multiple_ `CounterGather` objects from multiple databases).

Below is some demo code implementing multi-database gather; it is essentially what is used by `_find_best` in `src/sourmash/search.py`.

```python
    def gather(self, query, threshold_bp):
        results = []

        best_result = None
        best_intersect_mh = None

        # find the best score across multiple counters, without consuming       
        for counter in self.counters:
            result = counter.peek(query.minhash, query.minhash.scaled,
                                  threshold_bp)
            if result:
                (sr, intersect_mh) = result

                if best_result is None or sr.score > best_result.score:
                    best_result = sr
                    best_intersect_mh = intersect_mh

        if best_result:
            # remove the best result from each counter                          
            for counter in self.counters:
                counter.consume(best_intersect_mh)

            # and done!                                                         
            return [best_result]
        return []
```

## Implementation checklist etc.

TODO:

- [x] add a `prefetch --linear` to force a _linear_ pass across databases, e.g. for the current large GenBank SBTs.
- [x] provide a test fixture so we can run all the gather tests with and without `--prefetch`
- [x] clean up and comment CounterGather class (preparatory to #1489, tho)
- [x] address https://github.com/dib-lab/sourmash/issues/1220 with iterative output?
- [x] write command line docs
- [x] add `--save-prefetch` to `sourmash gather`
- [x] test internal `Index.prefetch` method at Python level
- [x] figure out how to test the laziness and the streaming
- [x] consider extracting the `SaveSignaturesToLocation` code to a new PR.
- [x] update PR description to include #1370, etc.
- [x] test LCA database check code (moved to https://github.com/dib-lab/sourmash/pull/1495)
- [x] revamp tests for prefetch and linear using new fixtures
- [x] test `LazyLinearIndex` for laziness!
- [x] test `LazyLinearIndex` bool and len
- [x] test `ZipFileLinearIndex` bool
- [x] test `ValueError` catch for prefetch_databases

### CODE REVIEW

**Comment by ctb** [comment]
Update: merged into this PR already.

----

Some more thoughts now that I've gotten this first pass implemented -

## make prefetch a method on `Index`, maybe?

this fits with @luizirber [dishwasher thinking re greyhound](https://github.com/dib-lab/sourmash/issues/1226#issuecomment-720478514) and the 'counter_gather' method in https://github.com/dib-lab/sourmash/pull/1311/, but I would suggest instead having the `prefetch` method return another `Index` object that can be used for gather or search.

i.e. from [this comment on the greyhound issue](https://github.com/dib-lab/sourmash/issues/1226#issuecomment-720815108),
>For step 1, the query is loaded in memory and then all the reference sigs are loaded in parallel and checked for intersection size with the query (map). If they are above the threshold_bp, the hashes are added to a LCA-like DB (reduce). A Counter-like (in the Python sense) structure is built from the DB, for ordering references by intersection size with the query.

>For step 2, while there are matches in the counter the top match is selected, and each reference in the counter has their count decreased if they have hashes in the top match (using the revindex to find which references to decrease).

I would have step 1 return an `Index` data structure to support step 2. Then we'd use the normal gather code for everything else.

**note: this is started in #1371**

**Comment by ctb** [comment]
note to self: `prefetch` only supports scaled signatures.

**Comment by ctb** [comment]
note to self: prefetch should _explicitly_ support streaming, single pass across database, underneath.

this is mildly trick to manage, and even harder to test...

**Comment by ctb** [comment]
Notes to self:

* review with an eye to supporting streaming & lazy evaluation
* provide a `--linear` flag somewhere in here that switches from `load_file_as_index` to `load_file_as_signatures`

**Comment by ctb** [comment]
Note to self: avoid collecting all the signatures in memory!

**Comment by ctb** [comment]
random thought: can we combine or share code between CounterGather and LCA database?

**Comment by ctb** [comment]
Inquiries for @luizirber -

## Oxidizing prefetch

The method signature for `Index.prefetch` is:
```
     def prefetch(self, query, threshold_bp, **kwargs):
```
and it is a generator function that yields signatures with >= `threshold_bp / query.minhash.scaled` overlap, in no particular order. I feel like this is the most likely target for oxidation. Am I correct? :) If so, is it at the right level of abstraction? I don't believe any of the Index subclasses are oxidized yet.

## Oxidizing `CounterGatherIndex`

`CounterGatherIndex` is the class that holds prefetch results for a particular query and then does a revindex-based `gather`. Is this something else that you would see as an oxidation target? If so, what kind of API would best suit?

**Comment by ctb** [comment]
update - extracted this to https://github.com/dib-lab/sourmash/pull/1493.

---

channeling @bluegenes request about a more flexible `--save-matches` https://github.com/dib-lab/sourmash/issues/1440, I implemented a feature in `sourmash prefetch` for which I welcome feedback -

if `sourmash prefetch` is given a `--save-matches` with a trailing `/`, it saves matches as files in that directory.

for example,
```
sourmash prefetch -k 31 tests/test-data/47.fa.sig tests/test-data/63.fa.sig  --save-matches=bar/
```
saves the matching signature as `bar/38729c63.sig`.

In particular, this lets us avoid keeping matching signatures in memory, which is extra nice for large databases / long-running queries; ref better UX for massive collections https://github.com/dib-lab/sourmash/issues/1350.

- [x] consider compressing the output signatures by default

**Comment by luizirber** [comment]
> and it is a generator function that yields signatures with >= `threshold_bp / query.minhash.scaled` overlap, in no particular order. I feel like this is the most likely target for oxidation. Am I correct? :) If so, is it at the right level of abstraction? I don't believe any of the Index subclasses are oxidized yet.

The issue with this approach is "serializing" the execution: because the generator/yielding is driving the computation, it not only returns values one at a time, but end up forcing calculating values one at a time. This is good for streaming and keeping low memory consumption, but doesn't really explore the parallelism of the situation.

So, oxidizing this bit might be faster, but probably not much (or can even be slower depending on the overhead of sending data across FFI).

The way that prefetching was implemented in https://github.com/luizirber/2020-11-02-greyhound is by batching (give it a pile of paths to signatures, get back a smaller pile of sigs/paths), because it can avoid loading all sigs at once. But it is harder to do this with the prefetch method... (but still doable with `sourmash prefetch`, tho)

> ## Oxidizing `CounterGatherIndex`
> 
> `CounterGatherIndex` is the class that holds prefetch results for a particular query and then does a revindex-based `gather`. Is this something else that you would see as an oxidation target? If so, what kind of API would best suit?

I like the idea of it being something that takes other `Index` objects, but that makes it very hard to be oxidized (because how to we call back into Python code to grab results?). It's probably better for it to be Python-only. In #1238 I was going for a `.counter_gather()` method because that allows each `Index` to control how to generate the `Counter` (and that's easy to oxidize).

Can `CounterGatherIndex` be defined in terms of a `.counter_gather()` method on `Index`?

**Comment by ctb** [comment]
> > and it is a generator function that yields signatures with >= `threshold_bp / query.minhash.scaled` overlap, in no particular order. I feel like this is the most likely target for oxidation. Am I correct? :) If so, is it at the right level of abstraction? I don't believe any of the Index subclasses are oxidized yet.
> 
> The issue with this approach is "serializing" the execution: because the generator/yielding is driving the computation, it not only returns values one at a time, but end up forcing calculating values one at a time. This is good for streaming and keeping low memory consumption, but doesn't really explore the parallelism of the situation.
> 
> So, oxidizing this bit might be faster, but probably not much (or can even be slower depending on the overhead of sending data across FFI).
> 
> The way that prefetching was implemented in https://github.com/luizirber/2020-11-02-greyhound is by batching (give it a pile of paths to signatures, get back a smaller pile of sigs/paths), because it can avoid loading all sigs at once. But it is harder to do this with the prefetch method... (but still doable with `sourmash prefetch`, tho)

I guess we've been thinking at different abstraction levels, then; my ignorance of the details of the greyhound implementation is biting us.

I've been thinking about the oxidation layer like a bulky RPC mechanism, where we're trying to minimize calls and communication b/c the latency and format conversion is so high. I believe you're thinking at a more granular level. I have to think about this more before writing down anything concrete, but a hot take:

The prefetch method doesn't guarantee anything at all about order of traversal/returned signatures. So it should be completely compatible with parallelism and batching, no? And it sounds like the major concern you have with it is the
amount of information passing required - are we sending back thousands of _signatures_ via FFI, or what? I guess I don't have a good sense for how slow it is to pass around signatures; come to think of it you've expressed various concerns about the amount of hash copying etc etc that we're doing.

Hmm, maybe the passing back and forth of full URLs to signatures could be an underlying mechanism? That is, we implement specific prefetch methods for Zip files and SBTs (and eventually LCAs) that return URLs that can be understood by the Rust and Python layers?

> > ## Oxidizing `CounterGatherIndex`
> > `CounterGatherIndex` is the class that holds prefetch results for a particular query and then does a revindex-based `gather`. Is this something else that you would see as an oxidation target? If so, what kind of API would best suit?
> 
> I like the idea of it being something that takes other `Index` objects, but that makes it very hard to be oxidized (because how to we call back into Python code to grab results?). It's probably better for it to be Python-only. In #1238 I was going for a `.counter_gather()` method because that allows each `Index` to control how to generate the `Counter` (and that's easy to oxidize).

Yeah, I was worried about the Python callbacks.

Random idea, what do you think about oxidizing the `JaccardSearch` classes? There should only be a few of them, barring outre masking classes (e.g. `JaccardSearchBestOnly_ButIgnore` class in `test_index.py`), the logic is simple, and we could converge on a single Rust implementation that can be used by oxidized `Index` classes as well as the Python-based `Index` classes.

(Similarly, I feel like `LinearIndex` is ripe for oxidation. Maybe I'll put that in an issue.)

> Can `CounterGatherIndex` be defined in terms of a `.counter_gather()` method on `Index`?

The `CounterGatherIndex` API that I currently have in this PR smells bad. I like the idea of changing it up, will explore the `.counter_gather()` method idea.

**Comment by ctb** [comment]
> > I like the idea of it being something that takes other `Index` objects, but that makes it very hard to be oxidized (because how to we call back into Python code to grab results?). It's probably better for it to be Python-only. In #1238 I was going for a `.counter_gather()` method because that allows each `Index` to control how to generate the `Counter` (and that's easy to oxidize).
> > Can `CounterGatherIndex` be defined in terms of a `.counter_gather()` method on `Index`?
> 
> The `CounterGatherIndex` API that I currently have in this PR smells bad. I like the idea of changing it up, will explore the `.counter_gather()` method idea.

OK, I remembered a blocking aspect to this ;).

`gather` is a cross-database approach, and the `CounterGatherIndex` needs to be able to hold all the overlapping signatures from all the databases. So I'm not sure a `counter_gather()` method belongs on a single `Index` class?

Thinking about this as a [scatter-gather approach](https://www.enterpriseintegrationpatterns.com/patterns/messaging/BroadcastAggregate.html), I was envisioning `prefetch(...)` as a scatter-able method, and `CounterGatherIndex` as the gather method (apologies for name collision 😆 ).

Is there a way to provide an oxidized version of `CounterGatherIndex` that we can initialize with a query, and feed things into from Python, and also feed things into from Rust?

**Comment by ctb** [comment]
> > > I like the idea of it being something that takes other `Index` objects, but that makes it very hard to be oxidized (because how to we call back into Python code to grab results?). It's probably better for it to be Python-only. In #1238 I was going for a `.counter_gather()` method because that allows each `Index` to control how to generate the `Counter` (and that's easy to oxidize).
> > > Can `CounterGatherIndex` be defined in terms of a `.counter_gather()` method on `Index`?
> > 
> > 
> > The `CounterGatherIndex` API that I currently have in this PR smells bad. I like the idea of changing it up, will explore the `.counter_gather()` method idea.
> 
> OK, I remembered a blocking aspect to this ;).
> 
> `gather` is a cross-database approach, and the `CounterGatherIndex` needs to be able to hold all the overlapping signatures from all the databases. So I'm not sure a `counter_gather()` method belongs on a single `Index` class?
> 
> Thinking about this as a [scatter-gather approach](https://www.enterpriseintegrationpatterns.com/patterns/messaging/BroadcastAggregate.html), I was envisioning `prefetch(...)` as a scatter-able method, and `CounterGatherIndex` as the gather method (apologies for name collision 😆 ).
> 
> Is there a way to provide an oxidized version of `CounterGatherIndex` that we can initialize with a query, and feed things into from Python, and also feed things into from Rust?

OK, in https://github.com/dib-lab/sourmash/pull/1489, I have a reasonably clean refactoring of the `CounterGather` into something that could work on individual `Index` classes. It could be what is returned from a `.counter_gather()` method, for example. Curious as to your thoughts, @luizirber!

UPDATE: implemented `counter_gather` etc etc.

**Comment by ctb** [comment]
- [ ] `--save-prefetch` to `sourmash gather`

**Comment by ctb** [comment]
so @luizirber merged #1489 before I got a chance to reflect that the same performance regression he identified in https://github.com/dib-lab/sourmash/pull/1392#issuecomment-822767431 - visiting too many nodes! - is now institutionalized in #1489 and now this PR!  (We [figure it out](https://github.com/dib-lab/sourmash/pull/1474#issuecomment-824362643) back then, of course.)

But I thought about it yesterday, and - hot take - I think I will need to revisit gather on `Index` so that we provide the `peek` method appropriately, and provide `consume` as a no-op.  That is, we need to implement the `CounterGather` iteration interface on `Index`. I suspect that the current `Index.gather` is basically what we need, but will confirm.

**Comment by luizirber** [comment]
> so @luizirber merged #1489 before I got a chance to reflect that the same performance regression he identified in [#1392 (comment)](https://github.com/dib-lab/sourmash/pull/1392#issuecomment-822767431) - visiting too many nodes! - is now institutionalized in #1489 and now this PR! (We [figure it out](https://github.com/dib-lab/sourmash/pull/1474#issuecomment-824362643) back then, of course.)

It is faster, but is using more memory (SBT index):
image

**Comment by luizirber** [comment]
The missing optimization for this one is calling `.unload()` for internal nodes in `prefetch` for SBTs, it is the same case as `search` (we only see the data once, so we can unload it)

**Comment by ctb** [comment]
I was thinking more about tree truncation, which isn't done by the current gather code. Unload is also important, tho. Complexities!

Still the part where this is faster doesn't make sense to me 😕 . Regardless, I bow before the benchmarks!

**Comment by ctb** [comment]
(IT'S LITERALLY TWICE AS FAST. WHAT?!)

**Comment by luizirber** [comment]
> (IT'S LITERALLY TWICE AS FAST. WHAT?!)

I think it makes sense, since we are replacing the iterative (and expensive) multiple best-match searches on SBTs (which depends on how comples the query is) with a linear pass (prefetch) and cheap iterative process (counter-gather). Even unoptimized it is likely to be way faster than the previous way of doing things!

> I was thinking more about tree truncation, which isn't done by the current gather code. Unload is also important, tho. Complexities!

Tree truncation will shine when
- the SBT is localized (#925 #1201), because we can prune it even earlier than what we can do today
- with extremely large trees, where even loading all sigs is expensive. If we can avoid searching large chunks of the tree we also avoid loading many sigs from disk, so it can be very low-memory compare to revindex.

I think always unloading on `gather` now is fine, since it will trigger another signature load for the matches, but will always be correct: data that is not persistent on disk/loadable from an `Storage` is not unloaded with the current code, so we don't risk losing the data.

**Comment by ctb** [comment]
jn Sun, May 02, 2021 at 04:46:45PM -0700, Luiz Irber wrote:
> > (IT'S LITERALLY TWICE AS FAST. WHAT?!)
> 
> I think it makes sense, since we are replacing the iterative (and expensive) multiple best-match searches on SBTs (which depends on how comples the query is) with a linear pass (prefetch) and cheap iterative process (counter-gather). Even unoptimized it is likely to be way faster than the previous way of doing things!

yar.

> > I was thinking more about tree truncation, which isn't done by the current gather code. Unload is also important, tho. Complexities!
> 
> Tree truncation will shine when
> - the SBT is localized (#925 #1201), because we can prune it even earlier than what we can do today
> - with extremely large trees, where even loading all sigs is expensive. If we can avoid searching large chunks of the tree we also avoid loading many sigs from disk, so it can be very low-memory compare to revindex.
> 
> I think always unloading on `gather` now is fine, since it will trigger another signature load for the matches, but will always be correct: data that is not persistent on disk/loadable from an `Storage` is not unloaded with the current code, so we don't risk losing the data.

And I think low memory gather can be done with prefetch first, then a separate
gather process.

--titus

**Comment by ctb** [comment]
OK, this rather monstrous PR is ready for review! @bluegenes @luizirber

File: doc/command-line.md [hunk 1]
```diff
@@ -57,16 +57,17 @@ species, while the third is from a completely different genus.
 
 To get a list of subcommands, run `sourmash` without any arguments.
 
-There are six main subcommands: `sketch`, `compare`, `plot`,
-`search`, `gather`, and `index`.  See [the tutorial](tutorials.md) for a
-walkthrough of these commands.
+There are seven main subcommands: `sketch`, `compare`, `plot`,
+`search`, `gather`, `index`, and prefetch.  See
```
Comment (bluegenes) [inline doc/command-line.md]:
```suggestion
`search`, `gather`, `index`, and `prefetch`.  See
```

File: doc/command-line.md [hunk 2]
```diff
@@ -326,6 +350,54 @@ containing a list of file names to index; you can also provide individual
...
+"streaming mode" on its inputs, in that each input file is loaded,
+searched, and then unloaded.  And `sourmash prefetch` can be run
+separately on multiple databases, after which the results can be
+searched in combination with `search`, `gather`, `compare`, etc.
+
+A motivating use case for `sourmash prefetch` is to run it on multiple
+large databases with a metagenome query using `--threshold-bp=0`,
+`--save-matching-hashes matching_hashes.sig`, and `--save-matches
+db-matches.sig`, and then run `sourmash gather matching-hashes.sig
+db-matches.sig`.
```
Comment (bluegenes) [inline doc/command-line.md]:
```suggestion
A motivating use case for `sourmash prefetch` is to run it on multiple
large databases with a metagenome query using `--threshold-bp=0`,
`--save-matching-hashes matching_hashes.sig`, and `--save-matches
db-matches.sig`, and then run `sourmash gather matching-hashes.sig
db-matches.sig`. 

This combination of commands ensures that the more time- and memory-intensive `gather`, step is run only on the smallest possible number of signatures without affecting the results.
```
...or similar? Just trying to drive the point home

**Review by bluegenes — COMMENTED** [review]
wow, this pr is a beast! I shall be back with more review comments.

File: doc/command-line.md (hunk 2 above)
Comment (ctb) [reply doc/command-line.md]: yes! thanks :)

**Comment by luizirber** [comment]
Going thru it, but while I check the code:

We let something slip in a previous PR (probably https://github.com/dib-lab/sourmash/pull/1392 or https://github.com/dib-lab/sourmash/pull/1477?) which made `search` slower:
image

and this PR still has the memory consumption problem in `gather` (same as https://github.com/dib-lab/sourmash/pull/1370#issuecomment-830834270)

**Comment by ctb** [comment]
> Going thru it, but while I check the code:
> 
> We let something slip in a previous PR (probably #1392 or #1477?) which made `search` slower:
> image

huh, ok.
 
> and this PR still has the memory consumption problem in `gather` (same as [#1370 (comment)](https://github.com/dib-lab/sourmash/pull/1370#issuecomment-830834270))

Could you try that with `--no-prefetch` please? and maybe `--linear` with or without `--prefetch`?

**Comment by ctb** [comment]
One random thought on the memory stuff - I don't think we're passing in `unload_data` to the SBT, maybe that would do it?

**Comment by luizirber** [comment]
> One random thought on the memory stuff - I don't think we're passing in `unload_data` to the SBT, maybe that would do it?

Yeah, I used the hammer in https://github.com/dib-lab/sourmash/compare/add/prefetch_cli...add/prefetch_cli_unload and set unload to `True` as default, will report soon.

**Comment by luizirber** [comment]
HAH, THIS IS GLORIOUS

image

**Comment by ctb** [comment]
@luizirber suggested adding some information to `sourmash prefetch` and (presumably) `gather --prefetch` about how many signatures have been collected, and what input file it's on.

**Comment by ctb** [comment]
another thought for changes - here or in another PR - is to make sure that the top-level APIs for search and Index are using `SourmashSignature` and not `MinHash` objects. That protects us from specific details of the sketch, and potentially lets us make use of fine tuned, or more advanced, or just plain different sketch types.

Right now most of the top-level APIs do use `query` rather than `query_mh`, but the `peek` and `consume` methods and the `make_gather_query` function take MinHash directly. Just something to think about.

**Comment by luizirber** [comment]
> another thought for changes - here or in another PR - is to make sure that the top-level APIs for search and Index are using `SourmashSignature` and not `MinHash` objects. That protects us from specific details of the sketch, and potentially lets us make use of fine tuned, or more advanced, or just plain different sketch types.
> 
> Right now most of the top-level APIs do use `query` rather than `query_mh`, but the `peek` and `consume` methods and the `make_gather_query` function take MinHash directly. Just something to think about.

:+1: :+1: 

(but I think punt to another issue, this is already too long...)

EDIT: punted to https://github.com/dib-lab/sourmash/issues/1514

**Comment by ctb** [comment]
> @luizirber suggested adding some information to `sourmash prefetch` and (presumably) `gather --prefetch` about how many signatures have been collected, and what input file it's on.

this is actually there --
```
loading signatures from 'gtdb-r202.genomic-reps.k31.zip'
total of 78 matching signatures.
```
but it relies on having actual matches. I'm not sure how to do a progress indicator given that `Index.find` doesn't provide any information about how many signatures have been searched.

**Review by luizirber — APPROVED** [review]

**Comment by ctb** [comment]
> > One random thought on the memory stuff - I don't think we're passing in `unload_data` to the SBT, maybe that would do it?
> 
> Yeah, I used the hammer in [add/prefetch_cli...add/prefetch_cli_unload](https://github.com/dib-lab/sourmash/compare/add/prefetch_cli...add/prefetch_cli_unload) and set unload to `True` as default, will report soon.

thanks for merging! did we want to add this unload branch in?

**Comment by luizirber** [comment]
> thanks for merging! did we want to add this unload branch in?

-> https://github.com/dib-lab/sourmash/pull/1513

**Pull request merged**